core/src/main/scala/chisel3/Bits.scala

```diff
@@ -148,8 +148,22 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
 
   /** @group SourceInfoTransformMacro */
   final def do_extract(x: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool = {
+    this.widthOption.foreach { thisWidth =>
+      if (thisWidth == 0) {
+        Builder.error(s"Cannot extract from zero-width $this")
+      } else {
+        x.widthOption.foreach { xWidth =>
+          if (xWidth >= 31 || (1 << (xWidth - 1)) >= thisWidth) {
+            Builder.warning(s"Dynamic index '$x' is too large for extractee '$this'")
```

Comment:
```suggestion
            Builder.warning(s"Dynamic index with width $xWidth is too large for extractee of width $thisWidth")
```
In the interest of having these warning deduplicate, let's go back to your original message. I'm working on an improvement to all Chisel error reporting (having the source line and a caret) so having that along with these messages will be pretty good.